### DISCUSSION TITLE
Code coverage report running locally

### DISCUSSION THREAD

**Discussion by s-KaiNet** [pnp/pnpcore#603] [Ideas] [answered]
It's not an issue, it's more a question, thus opening a new discussion. One of the "dev" goals for this library is to maintain good code quality (including, but not limited to test code coverage >80%).

As a contributor, I haven't found an easy way to see the coverage for my code (total coverage, but what is most importantly line coverage to see what is covered and what is not). Maybe I haven't searched well, but I ended up with a custom PS script, which can run tests for a specific library or even a specific class. It opens coverage report in a browser (HTML format) so that you can see actual line coverage for any file. Under the hood, the script uses `dotnet test` command and the [ReportGenerator](https://github.com/danielpalme/ReportGenerator) tool. 

<details>
  <summary>Here is the code I use</summary>

```powershell
  <#
  .SYNOPSIS
	  The function runs unit tests for the selected library or class and prints the code coverage report using html format.
	  
  ​.PARAMETER TestProjectName
	  Test project name, for example PnP.Core.Test. Either TestProjectName or FqdnClassName should be provided.
  
  ​.PARAMETER FqdnClassName
	  Fully qualified class name, for example PnP.Core.Test.SharePoint.FilesTests. Either TestProjectName or FqdnClassName should be provided.
	  
  ​.PARAMETER OpenReport
	  Whether to open a report using the default browser, default is TRUE
  
  .EXAMPLE
	  .\CoveragePnPCore.ps1 -FqdnClassName PnP.Core.Test.SharePoint.FilesTests
  ​
	  The above command will run all tests in the specified class and will launch a browser with the coverage report
	  
	  .\CoveragePnPCore.ps1 -TestProjectName PnP.Core.Test -OpenReport $false
	  
	  The above command will run all tests in the specified test project and will NOT open a browser with the coverage report
  #>
  param (
  [Parameter(Mandatory = $false)]
  [string]$TestProjectName,
  
  [Parameter(Mandatory = $false)]
  [string]$FqdnClassName,
  
  [Parameter(Mandatory = $false)]
  [bool]$OpenReport = $true)
  
  if(!$TestProjectName -and !$FqdnClassName) {
	  throw "You should provide either -TestProjectName or -FqdnClassName parameter"
	  return;
  }
  
  $projectPath = ""
  
  if($TestProjectName) {
	  $projectPath = "pnpcore/src/sdk/$($TestProjectName)/$($TestProjectName).csproj"
  } else {
	  $index = $FqdnClassName.IndexOf(".Test.") + ".Test.".Length;
	  $libName = $FqdnClassName.Substring(0, $index - 1)
	  $projectPath = "pnpcore/src/sdk/$($libName)/$($libName).csproj"
  }
  
  $tempFolder = $env:TEMP
  $pnpTestsFolder = "$($tempFolder)/pnp-sdk-tests"
  
  $testResultsPath = "$($pnpTestsFolder)/TestResults"
  $coveragePath = "$($pnpTestsFolder)/coverage"
  
  if (Test-Path -Path $pnpTestsFolder) {
	  Remove-Item -Recurse -Force $pnpTestsFolder
  }
  
  
  if(!$FqdnClassName) {
	  dotnet test $projectPath --no-restore --verbosity normal --collect:"XPlat Code Coverage" --results-directory $testResultsPath
  } else {
	  dotnet test $projectPath --no-restore --verbosity normal --collect:"XPlat Code Coverage" --results-directory $testResultsPath --filter ClassName=$FqdnClassName
  }
  
  & $env:USERPROFILE\.dotnet\tools\reportgenerator.exe -reports:"$($testResultsPath)/**/*.cobertura.xml" -targetdir:$coveragePath -reporttypes:html
  
  $coverageHtml = "$($coveragePath)/index.html"
  
  Write-Host ""
  Write-Host "The coverage file is generated under '$($coverageHtml)'" -ForegroundColor Yellow
  Write-Host ""
  
  if($OpenReport) {
	  Invoke-Item $coverageHtml
  }
```  

</details>

So you can run 

```powershell
.\CoveragePnPCore.ps1 -FqdnClassName PnP.Core.Test.SharePoint.FilesTests
```
to see coverage for individual test class or the whole library:

```powershell
.\CoveragePnPCore.ps1 -TestProjectName PnP.Core.Test -OpenReport $false
```

This is the sample output from a file coverage where you can click and see the line coverage for a file:

![image](https://user-images.githubusercontent.com/10658276/139478637-6e894d2d-11e5-447d-9124-8252e7645adc.png)

Does it make sense to include this script in the repo? It's IDE independent, however runs only on Windows (because of `reportgenerator.exe`). If yes, I will include it and update documentation for contributors.

**Comment by jansenbe** [comment, marked as answer]
Great suggestion @s-KaiNet!. Visual Studio Enterprise has built in code coverage tools (see https://visualstudio.microsoft.com/vs/compare/ for version comparison), but not every developer has access to this version. Having an alternative approach is great for folks using VSCode or other IDE's. Would be great if you could add your script under `src/tools` and add relevant documentation in `docs/contributing/writing%20tests.md`. 

Thanks for helping out!

FYI: VS 2022 experience:
![image](https://user-images.githubusercontent.com/7451219/139527162-1f6aa839-7aff-4594-929a-52781a5eae6c.png)
